### ISSUE TITLE
Spike Timing loss/accuracy functions in snn.functional

### ISSUE THREAD

**Issue by jeshraghian** [jeshraghian/snntorch#80]
Other than `fucntional.mse_membrane_loss()`, all other loss functions available are only able to train rate-coded networks.

Temporal / time-to-first-spike mechanisms are more challenging as they involve a conversion from the spike (1/0) domain to the time domain. The most obvious fix is to simply iterate through the output spike tensor's time-axis, and populate a new tensor with the time at which each spike occurs. This would then be passed into a MSELoss function.

But indexing breaks the computational graph in PyTorch, so this would have to be overridden using a pre-defined `backward` function that is inherited from `torch.autograd.Function`.

Bonus features: 
* tolerance for spike-timing
* relative spike-timing
* account for all spikes, not just the first

**Comment by jeshraghian** [comment]
Done - scheduled for release in 0.5.0.
